chore: merge branch mergeback_chore-release_8.4.0 into edge

smb2268 opened the pull request:
Merge chore_release-8.4.0 mergeback branch into edge

y3rsh committed:
test(api): liquid classes blowout properties (#17875)

mjhuff committed:
fix(app): Fix blocking Flex LPC for OT-2 reasons (#17890)

It's possible for users to perform LPC on the Flex without a tiprack in the protocol, and we should, you know, let users LPC, since that's really just an OT-2 restriction. A minor CSS fix is included here too for the OT-2 wizard header.

ryanthecoder committed:
feat(hardware, api): Add a python-controlled tattle flag for resin tip dispensing (#17851)

We count every single evotips_dispense that happens on a pipette and store it in the eeprom. We report it when you connect the pipette, and we can add more structured data from it if you want.

sanni-t committed:
feat(shared-data): enable multi dispense blowout (#17895)

Closes RQA-3980

# Overview

Updates all liquid classes' multi-dispense blowout entries to:
- specify blowout location of `trash`
- specify blowout flow rate to be the same as the single dispense flow
rate specified for that pipette+tip combo
- enable blowout

## Risk assessment

Low. Just enables a setting that was disabled by default. We already
have tests that check that transfers happen correctly when these
properties are enabled

mjhuff committed:
fix(app): Fix LPC not POSTing labware defs to the maintenance run (#17894)

Closes RQA-4010

Currently, LPC utilizes the existing labware definitions from runs/:runId/loaded_labware_defintions and POSTs those definitions to the LPC maintenance run. However, it turns out that those definitions do not necessarily exist on the run resource at the time of LPC maintenance run creation.

We actually have access to the labware definitions from analysis directly and don't need to query the server for them. A much simpler (and functional) approach is to just POST the definitions we already have.

andySigler committed:
fix(api): Estimating volume when liquid height is zero should return zero not error (#17909)

(copy of other PR 17877 that was merged into edge earlier)

As written before, the _find_volume_in_partial_frustum function would raise an error if the target_height equaled the top or bottom of any frustum. For example, 0.0 in my case, because I was getting this error during my protocol run:

opentrons.protocol_engine.errors.exceptions.InvalidLiquidHeightFound: Error 4000 GENERAL_ERROR
(InvalidLiquidHeightFound): Unable to find volume at given well-height 0.0.

sfoster1 committed:
chore(api,shared-data): update evotip definitions (#17907)

Update the labware definitions for the evotip adapters. These are the
existing geometry for the 4 v2 adapters; they're getting new names (and
thus going to v1), and they're getting their branding details filled in.
We've removed the v1 adapters entirely, since their names were
confusingly close to the new ones.

This also adds the adapters to the labware tab and to the labware
library. Note that it does _not_ add them to PD; PD can't create evotip
protocols, and anyway we only want people to be using these with one
specific protocol.



Closes EXEC-1378
Closes EXEC-1364

mjhuff committed:
fix(app): fix not applying offsets for locations containing modules (#17901)

Closes RQA-4012

Fixes an issue applying offsets when the offsetLocationSequence involves a module. The util that converts the locationSequence to an offsetLocationSequence assumed that the ordering of the sequence should translate 1:1, but for offsetLocationSequence, the addressableArea kind component must be last in order to successfully POST updates for the offset.

mjhuff committed:
refactor(app): Add OT-2 LPC "skip" and immediately apply offsets (#17902)

Closes EXEC-1360

Adds the scoped LPC Redesign functionality for the OT-2:

Skipping! You can now choose not to LPC a tip rack or a labware. You still can't skip the "attach tip" and "detach tip" step for obvious reasons, though.

Applying offsets after each step! Yes, exiting LPC early means you do get to keep the offset data that you set or previously existed.

emilyburghardt committed:
chore: 8.4.0 alpha release notes (#17892)

<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure
you've read the "Opening Pull Requests" section of our Contributing
Guide:


https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams,
pictures, and videos along with text formatting make it possible to
create a rich and informative PR. For more information on GitHub
markdown, see:


https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out
the sections below to the best of your ability!
-->

# Overview

<!--
Describe your PR at a high level. State acceptance criteria and how this
PR fits into other work. Link issues, PRs, and other relevant resources.
-->

Release notes for 8.4.0-alpha. 

## Test Plan and Hands on Testing

<!--
Describe your testing of the PR. Emphasize testing not reflected in the
code. Attach protocols, logs, screenshots and any other assets that
support your testing.
-->

## Changelog

<!--
List changes introduced by this PR considering future developers and the
end user. Give careful thought and clear documentation to breaking
changes.
-->

Added new features, improvements, and bug fixed (including [8.3.0 punted
bugs](https://opentrons.atlassian.net/browse/RQA-3897))

## Review requests

<!--
- What do you need from reviewers to feel confident this PR is ready to
merge?
- Ask questions.
-->

-any changes to liquid classes customer messaging? 
-any missing bug fixes or improvements? 

## Risk assessment

<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your
change may affect some other part of the system.
- For instance, changing return tip behavior may also change the
behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's
risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is
enough.
-->
low.

mjhuff committed:
fix(app): Fix ODD LPC labware selection (#17913)

Closes RQA-4014

On the ODD during LPC, on the Labware List view, if you click anything but the first labware, you will still select the first labware in the list. This PR makes it so when you click on not the first labware, you get the labware you actually selected.

mjhuff committed:
fix(app): Fix LPC offset tables not in sync (#17917)

loses RQA-4013

There are a couple pieces of LPC state that are not necessarily controlled by the active app: the deck config and the labware info (which contains offset data). I recently moved the logic for updating store changes whenever the deck config changed to a separate hook, but I didn't do the same for the labware info, woops. This means that if a diferent user was actively performing LPC, offset updates wouldn't reflect in protocol setup offset table.

This commit creates granular state update actions for both the deck config and labware info.

mjhuff committed:
fix(app): allow LPCing default offsets for 'stackingOnly' labware  in an adapter (#17918)

Closes EXEC-1390

Certain labware should always be LPC'd in an adapter, even when setting a default offset (eg, evotips). To determine whether or not we require an adapter for all offset calculations, we examine the stackingOnly quirk present on the labware definition, and add the necessary info to the default offset.

SyntaxColoring committed:
refactor(app): Drive device reset settings fully from client (#17921)

## Overview

This refactors the desktop app and ODD components for doing a device
reset.

Goes towards EXEC-1281.

## Background

Historically, all of the reset options were server-defined. The client
would fetch the available options from the server (via `GET
/settings/reset/options`) and display them as-is, with server-defined
text.

More recently, that model has been stretched beyond its limits:

* We want the UI to be localized with frontend-defined strings.
* We want frontend to control the options' ordering and grouping.
* We want some reset options, like `onDeviceDisplay`, to be special:
they should be hidden from the UI and selected implicitly depending on
the user's selection of other options.
* We want the server to implement some of these reset options through
separate endpoints, instead of all of them through `/settings/reset`.
(See e.g. [this
discussion](https://github.com/opentrons/opentrons/pull/17051#discussion_r1872275665),
though I sometimes have second thoughts about this.)

So we have been in a weird in-between where we were getting the dynamic
list of options from the server, but still making strong assumptions
about what that list would contain. This PR rewrites it to be fully
client-decided.

## Changelog

This is intended to not have any user-facing change.

The one "accidental" change that I'm aware of is that if the client is
newer than the server and has an extra reset option, it will now be
displayed, and filtered out of the HTTP request. Formerly, it would not
even be displayed.

sanni-t committed:
fix(shared-data): fix multi dispense blowout flow rate values (#17904)

# Overview

Realized a few wrong values in my previous PR #17895. 
The multi-disp blowout flow rates are supposed to be the max of the
single dispense flow rates for that pipette+tiprack. A couple of entries
got min values instead. Fixed those

## Risk assessment

Low. Small fixes in definitions

sanni-t committed:
feat(api): add tip returning to distribute_liquid (#17920)

Closes AUTH-1496

# Overview

Implements the tip returning functionality to `distribute_liquid`.

## Risk assessment

Medium-Low. Affects `distribute_liquid` tip handling, but has decent
test coverage to avoid accidental behavior changes.

smb2268 committed:
feat(app): Refactor labware details into labware & liquid details, accomodate lid stacking (#17896)

Fix EXEC-1220, EXEC-1223, EXEC-1303, EXEC-1304, EXEC-1305, EXEC-1092

jbleon95 committed:
feat(api): group together multiple wells for multi-channel transfers (#17900)

adds multi-channel pipette support to the liquid class transfer methods (`transfer_liquid`, `distribute_liquid`, `consolidate_liquid`) by means of adding well grouping logic for multi-tip configurations addressing 96 and 384 well plates.

mjhuff committed:
fix(app): Do not LPC in staging area slots (#17926)

Closes RQA-4016

Believe it or not, we shouldn't LPC in staging area slots. To fix, we just need to mark LPC-able commands as being "not valid" if they are in a staging slot addressable area.

sfoster1 committed:
chore(shared-data): show tiprack lid in LL (#17925)

What it says on the tin, removes the tiprack lid labware from the
blocklist. You don't buy this on its own so it doesn't have branding
details and a sku.

mjhuff committed:
feat(app): Add default offset info banner to LPC (#17923)

Closes EXEC-1392 and RQA-4018

SyntaxColoring committed:
feat(app): Add an option to clear labware offsets (#17929)

sfoster1 committed:
feat(api): prepare for aspirate in air gap (#17919)

We should add a prepare-to-aspirate to the python protocol api `air_gap`
command, where it can issue the command after we've moved to a known-dry
position and before we actually dispense the airgap command. We know the
position is dry and therefore it's safe to do a prepare.

However, to be able to _unconditionally_ do a prepare, we also need to
add a fix to the engine prepare command to no-op if the pipette is
already prepared to aspirate. This is pretty important because if we've
already aspirated some liquid, we really don't want to move back to the
bottom position and dispense it all.


Useful testing protocol:
```
from opentrons.protocol_api import ProtocolContext

metadata = {"protocolName": "Test air gap prepare"}
requirements = {"apiLevel": "2.22", "robotType": "Flex"}


def run(protocol: ProtocolContext) -> None:
    plate = protocol.load_labware("biorad_384_wellplate_50ul", "A1")
    rack = protocol.load_labware("opentrons_flex_96_tiprack_50ul", "A2")
    pipette = protocol.load_instrument("flex_1channel_50", "left", tip_racks=[rack])

    pipette.pick_up_tip()
    pipette.aspirate(location=plate["A1"].bottom(z=2), volume=10)
    pipette.dispense(location=plate["A1"].top(), volume=10)
    pipette.blow_out(location=plate["A1"].top())
    pipette.air_gap(height=5, volume=10)

    pipette.dispense(location=plate["A2"].top(), volume=10)
    pipette.return_tip()
```


Closes EXEC-1391

sfoster1 committed:
fix(api): make the 8-channel evo pickup work (#17930)

These numbers can't be the same between press fit and cam action.
Whoops!

smb2268 committed:
Merge remote-tracking branch 'upstream/edge' into mergeback_chore_release-8.4.0

mjhuff reviewed (approved):
Thanks!

SyntaxColoring reviewed (approved):
LGTM with #17936 to fix the failing api test. Thank you!

y3rsh committed:
chore(shared-data): update command schema 13 (#17936)

This change that was made to command schema 12 - the command schema
releasing with Robot Stack v8.4.0 - needs to be added to the edge
current version schema of 13.

```shell
cd api
make command-schema
```

see #17889

smb2268 committed:
Merge branch 'edge' into mergeback_chore_release-8.4.0
